### DISCUSSION TITLE
Task-like type member access operator

### DISCUSSION THREAD

**Discussion by jmarolf** [dotnet/csharplang#5830] [Language Ideas]
## Summary
[summary]: #summary

Have a new operator that transforms task-like types to make them easier to work with in expressions.

## Motivation
[motivation]: #motivation

Method chaining is a very common way to manipulate data.

```csharp
void M(){
    string text = File.ReadAllText(path).ReplaceLineEndings();
}
```

and today expressions are one of the main ways in which data gets manipulated in the language.

Unfortunately, once async comes into the picture you need to wrap things in parenthesis in order for the associativity to keep working the way the developer wants _and_ we need to be in an async method.

```csharp
async void M(){ // need to be async method
    string text = (await File.ReadAllTextAsync(path)).ReplaceLineEndings(); // must wrap all awaits in parenthesis or have on separate line
}
```

You can still use `await` in expressions but due to the associativity mis-match it becomes harder to work with.

This is a similar problem that `?.` simplified for checking null:

```csharp
string? text = a == null ? null : a.ReplaceLineEndings();
```

```csharp
string? text = a?.ReplaceLineEndings();
```

If you are Stephen Toub you know that you can use `ContinueWith` to get the result from a task because the task has already completed

```csharp
string text = await File.ReadAllTextAsync(path).ContinueWith(static t => t.Result.ReplaceLineEndings());
```

But this is not obvious to some and doesn't make the expression easier to work with. There are also some 15+ overloads for `ContinueWith` for scenarios that are more advanced than what we are describing. In the same way that I can use `?.` to apply transformations over null values I want to easily do the same thing for task-like-types.

## Detailed design
[design]: #detailed-design

In other languages like C++ there were similar problems with pointer dereferencing. You had to first dereference the pointer and then call the member (`(*(e)).member`). For this reason C++ added the member access operator `->` so you could just write `e->member`.

```cpp
class Data
{
    public:
    int x;
};

int main()
{
    Data d;
    Data* ptr;
    ptr = &d

    // these two expressions are equivalent
    int x1 = (*(ptr)).x;
    int x2 = ptr->x;
}
```

I am proposing a similar syntactic concept:

```csharp
void M(){
    Task<string> text = File.ReadAllTextAsync(path)@.ReplaceLineEndings();
}
```

```antlr
task_like_type_member_access_expression
    : primary_expression null_conditional_operations
    ;

task_like_type_member_access_operations
    : task_like_type_member_access_operations@ '@' '.' identifier type_argument_list@
    | task_like_type_member_access_operations@ '@' '[' argument_list ']'
    | task_like_type_member_access_operations '.' identifier type_argument_list@
    | task_like_type_member_access_operations '[' argument_list ']'
    | task_like_type_member_access_operations '(' argument_list? ')'
    ;
```

Which would be semantically equivalent to

```csharp
void M(){
    Task<string> text = File.ReadAllTextAsync(path).ContinueWith(static t => t.Result.ReplaceLineEndings());
}
```

But the actual lowering could be something more efficient like this:

```csharp
void M(){
    Task<string> textTask = <>__asyncHelper(File.ReadAllTextAsync(path));
    static async Task<string> <>__asyncHelper(Task<string> task){
        string result = await task;
        result.ReplaceLineEndings();
        return result;
    }
}
```

I bring this lowered form up as a way to reason about what sets of constructs would be legal to use together but I think there are several valid approaches.

I should also note that I have no preference for `@` as the symbol that is used and invite the reader to replace it with whatever their heart tells them is right. Here are a few examples to help you out:

```csharp
Task<string> text = File.ReadAllTextAsync(path)|.ReplaceLineEndings();
```

```csharp
Task<string> text = File.ReadAllTextAsync(path)#.ReplaceLineEndings();
```

```csharp
Task<string> text = File.ReadAllTextAsync(path)~.ReplaceLineEndings();
```

```csharp
Task<string> text = File.ReadAllTextAsync(path)^.ReplaceLineEndings();
```

```csharp
Task<string> text = File.ReadAllTextAsync(path)%.ReplaceLineEndings();
```

```csharp
Task<string> text = File.ReadAllTextAsync(path)$.ReplaceLineEndings();
```

```csharp
Task<string> text = File.ReadAllTextAsync(path)`.ReplaceLineEndings();
```

Anyways, if we assume the compiler generates something semantically equivalent to this for these situations:

developer code:

```csharp
void M(){
    Task<string> text = File.ReadAllTextAsync(path)@.ReplaceLineEndings();
}
```

lowered:

```csharp
void M(){
    Task<string> textTask = <>__asyncHelper(File.ReadAllTextAsync(path));
    static async Task<string> <>__asyncHelper(Task<string> task){
        string result = await task;
        return result.ReplaceLineEndings();
    }
}
```

developer code:

```csharp
void M(){
    string oldValue = "foo";
    string newValue = "bar";
    Task<string> text = File.ReadAllTextAsync(path)@.Replace(oldValue, newValue);
}
```

lowered:

```csharp
void M(){
    string oldValue = "foo";
    string newValue = "bar";
    Task<string> textTask = <>__asyncHelper(File.ReadAllTextAsync(path), oldValue, newValue);
    static async Task<string> <>__asyncHelper(Task<string> task, string oldValue, string newValue){
        string result = await task;
        return result.Replace(oldValue, newValue);
    }
}
```

that implies that the following is not legal:

```csharp
class A {
    public void M(out int x);
}
void M(Task<A>) {
    Task ab = a@.M(out var x); 
}
```

Because it cannot be transformed in this manner:

```csharp
void M(Task<A> a) {
    Task ab = <>__asyncHelper(a, out var x) 
    static async Task <>__asyncHelper(Task<A> task, out int x){ // CS1988: Async methods cannot have ref, in or out parameters
        A result = await task;
        a.M(out x);
    }
}
```

However you could do something like this:

```csharp
void M(){
    Task task = new Task<List<string>>(() =>
    {
        List<string> result = new List<string>();
        result.Add("Hello");
        result.Add("World");
        return result;
    })@.Add("!");
}
```

which becomes

```csharp
void M(){
    Task task = <>__asyncHelper(new Task<List<string>>(() =>
    {
        List<string> result = new List<string>();
        result.Add("Hello");
        result.Add("World");
        return result;
    }));
    
    static async Task <>__asyncHelper(Task<List<string>>task){
        List<string> result = await task;
        result.Add("!");
    }
}
```

## Drawbacks
[drawbacks]: #drawbacks

The implicit assumption of this proposal is that monadic transformations over task-like types (this proposal in essence) are becoming as necessary as monadic transformations over `null` (`?.` and its family of `??` expressions). If this is not seen as necessary and we think that, while a special syntax to transform `null` makes sense in the language, nothing else in the type system needs this special treatment then this proposal doesn't really have any merit.

## Alternatives
[alternatives]: #alternatives

* [Forward pipe operators](https://github.com/dotnet/csharplang/discussions/74) would solve this and more.
* General monadic transformations in the language. While there is no proposal for this, having a general syntax for monadic transformations of a value in C# would solve this. If we think that is something we want to do we shouldn't special case task-like types and instead design the general case.
* Keep the same semantics of this design but instead have an agreed upon framework type that the compiler calls (perhaps just `ContinueWith`) instead of generating that static local function. This could also open the door up to allow developers to override what this operator does.

## Unresolved questions
[unresolved]: #unresolved-questions

This code

```csharp
void M(){
    string oldValue = "foo";
    string newValue = "bar";
    ConfiguredTaskAwaitable<string> text = File.ReadAllTextAsync(path)@.Replace(oldValue, newValue).ConfigureAwait(false);
}
```

Would seem to imply that we generate this

```csharp
void M(){
    string oldValue = "foo";
    string newValue = "bar";
    ConfiguredTaskAwaitable<string> textTask = <>__asyncHelper(File.ReadAllTextAsync(path), oldValue, newValue).ConfigureAwait(false);
    static async Task<string> <>__asyncHelper(Task<string> task, string oldValue, string newValue){
        string result = await task;
        return result.Replace(oldValue, newValue);
    }
}
```

The developer could always write:

```csharp
void M(){
    string oldValue = "foo";
    string newValue = "bar";
    ConfiguredTaskAwaitable<string> text = File.ReadAllTextAsync(path).ConfigureAwait(false)@.Replace(oldValue, newValue).ConfigureAwait(false);
}
```

to get

```csharp
void M(){
    string oldValue = "foo";
    string newValue = "bar";
    ConfiguredTaskAwaitable<string> textTask = <>__asyncHelper(File.ReadAllTextAsync(path).ConfigureAwait(false), oldValue, newValue).ConfigureAwait(false);
    static async Task<string> <>__asyncHelper(ConfiguredTaskAwaitable<string> task, string oldValue, string newValue){
        string result = await task;
        return result.Replace(oldValue, newValue);
    }
}
```

but its unclear if that is intuitive or helpful. I would _really_ like to avoid discussing context capture as part of this proposal but that may be something that we need to examine, or use as an argument to outright reject this.

**Comment by HaloFour** [comment]
Also, see: https://github.com/dotnet/csharplang/issues/4076

IMO, I'd much prefer a post-fix `await` to a sigil here.

**Reply by jmarolf** [reply]
The downside of that is that it could only be used inside async contexts. This could be used in any expression that matches the task-like type pattern.

**Reply by HaloFour** [reply]
Is that a common ask compared to wanting to just chain `await` operations within an expression?

**Reply by HaloFour** [reply]
Either way being a keyword vs. sigil didn't seem to be the determining factor as to why the LDM rejected that proposal.  If anything I'd think some members of the LDM would be even more opposed to a sigil shorthand form given discussions at wanting `await` boundaries to be relatively apparent.  But I'll leave that to the LDM to debate.  😄

**Reply by HaloFour** [reply]
Also: https://github.com/dotnet/csharplang/discussions/1117
https://github.com/dotnet/csharplang/discussions/827

**Reply by orthoxerox** [reply]
@jmarolf What expressions match the task-like type pattern, but *aren't* used inside async contexts?

**Reply by jmarolf** [reply]
@orthoxerox any method that returns a task like type but doesn't have an `async` keyword: https://grep.app/search?q=public%20Task But I only mention this to point out that its different from a postfix await operator in this subtle way. It certainly still related to https://github.com/dotnet/csharplang/issues/4076 and I think we would not want to do both. My intention was to try and scope this proposal a bit so we didn't get bogged down in implicit awaits too much but that may just be a distinction without a difference.


but I do think this is the central question:

1. Do we think this sort of thing is common enough that we need a special syntax for it? I've certainly needed it in the roslyn codebase and it appears based on @benaadams proposal from 2017 that there is some feedback about it https://github.com/dotnet/csharplang/discussions/1117 
2. Is this the right approach or is there an appetite for a more general design?

Only the LDM can really answer that so here we are.

**Reply by orthoxerox** [reply]
@jmarolf But that's just plain old Task, not some custom task-like, and task-like types don't even have to have a ContinueWith method: https://github.com/dotnet/roslyn/blob/main/docs/features/task-types.md. I understand what you were trying to say, though: one can call ContinueWith on a Task in any context, whereas await can appear only inside an async method.

**Comment by 333fred** [comment]
> General monadic transformations in the language. While there is no proposal for this, having a general syntax for monadic transformations of a value in C# would solve this. If we think that is something we want to do we shouldn't special case task-like types and instead design the general case.

As you know when we discussed this offline, this is the direction I'd like to go, if we want to go any direction here at all (I'm not 100% convinced of that yet). I'd much rather come up with a way to have a general bind operator that can apply to any type of monad, not just adding a single other one to C#'s quiver.

**Reply by jmarolf** [reply]
I also would be fine with this but wanted to discuss this because will take a while to get off the ground and a simple (relatively) translation like this would be easier to get through the LDM (again assuming we think any of this is a real problem to solve).

**Reply by orthoxerox** [reply]
What existing monadic-ish types do you suppose your operator should be designed to handle? `Nullable<T>`, `IEnumerable<T>`, `IQueryable<T>`, `Lazy<T>`, `Task<T>`?

**Reply by 333fred** [reply]
Any and all.

**Reply by sab39** [reply]
Is there an existing Issue or Discussion for this? It's obviously off-topic here but I'd love to read more about how it might work.

**Reply by jnm2** [reply]
Also hopefully it would be compatible with folks' custom types such as `Option<T>` and `Result<T>`.

**Reply by HaloFour** [reply]
So basically we're talking about shorthand for `SelectMany`/`FlatMap` operations, right?  Are we wading into `do`/`for` comprehension territory?

**Reply by 333fred** [reply]
> Is there an existing Issue or Discussion for this? It's obviously off-topic here but I'd love to read more about how it might work.

Almost certainly, but nothing that I've participated in.

**Reply by sab39** [reply]
I recall a blog series a while back (Eric Lippert maybe?) that was the first time I really got any kind of grasp on what monads are, which pointed out that to some degree they correspond to Linq operations. I remember that the Linq keywords don't really fit for most monads, but if you hand-wave away the problem of wrong keywords, how close does Linq come to being this kind of general mechanism, given appropriate extension methods for the types in question?

**Reply by HaloFour** [reply]
@sab39 

See: https://stackoverflow.com/a/4683716

LINQ is definitely much more geared to collections than to general purpose monads.

**Reply by 333fred** [reply]
I'm assuming you mean this series? https://ericlippert.com/2013/02/21/monads-part-one/

For another explanation, I'll let Brian Beckman do the talking: https://youtu.be/ZhuHCtR3xq8

**Reply by sab39** [reply]
Yes, that's the series I was talking about. (The fact that it was 9 years ago makes me feel old!)

I guess my question is: if we were to provide (a subset of) the same operations as Linq does, but with a different, more generalized, less collection-centric syntax, would that achieve the "general monadic transformations in the language" being discussed here, or would there also need to be some operations that Linq syntax can't express at all? And if the latter, what are those operations?

**Reply by jmarolf** [reply]
I'll let @333fred write his own proposal but as we were halving a hallway discussion about it you could think of it like linq and the query comprehension syntax. In the same way that a public method/extension method allows the `select from` statements to work You could have some general `Apply` function that would allow you to `apply`\`do`\ or something like that to an expression. Would potentially be an alternative to the pipes. Or at least I would not want us to do both pipes and this.

**Reply by HaloFour** [reply]
For reference:

[Haskell `do` notation](https://en.wikibooks.org/wiki/Haskell/do_notation)
[Scala `for` comprehensions](https://docs.scala-lang.org/tour/for-comprehensions.html) ([more detailed](https://www.baeldung.com/scala/for-comprehension))

**Reply by 333fred** [reply]
> I guess my question is: if we were to provide (a subset of) the same operations as Linq does, but with a different, more generalized, less collection-centric syntax, would that achieve the "general monadic transformations in the language" being discussed here, or would there also need to be some operations that Linq syntax can't express at all? And if the latter, what are those operations?

I don't know the answer :slightly_smiling_face:. I have not thought about this deeply, I just know I don't want another single monad bind operator in the language.

**Reply by jmarolf** [reply]
and I want _just one more Fred, pretty please?_ but am okay with us doing a more general feature. Probably not ok with the 5+ years it will take to do that but thats csharplang lyfe.

**Comment by Tinister** [comment]
```
string text = await File.ReadAllTextAsync(path).ContinueWith(static t => t.Result.ReplaceLineEndings());
```

Is this actually the recommended way of doing something like that?  My understanding was this creates a whole new task and does an extra schedule on the thread pool et al just to execute that synchronous lambda.

I wonder if a better strategy is to synthesize a lightweight TaskAwaiter-like struct that wraps the TaskAwaiter from earlier in the chain and performs the transformation as it bubbles out (e.g. `string GetResult() => wrappedAwaiter.GetResult().ReplaceLineEndings()`)

**Reply by HaloFour** [reply]
> My understanding was this creates a whole new task and does an extra schedule on the thread pool et al just to execute that synchronous lambda.

IIRC `TaskContinuationOptions.ExecuteSynchronously` will avoid re-scheduling on the thread pool.

**Reply by jmarolf** [reply]
`ContinueWith` was something I brought up because its an api that people could use today and what I am proposing has the same "semantics" or at least the use case for the `ContinueWith` api is the same. But this is certainly not the optimal code gen strategy. 

Yes, using an async local function (as I show later) would be a more efficient way to do this as the compiler would emit a struct for the async state machine (if nothing ever escaped the scope) and there doesn't even need to be any task created or allocated at all. I would need to think about what the _ideal_ emit strategy for this would be. No point in doing that until we agree on the general semantics.
